File: frameworks/motia/motia-js/packages/motia/src/multi-trigger.ts
import type {
  ExtractApiInput,
  ExtractQueueInput,
  ExtractStateInput,
  ExtractStreamInput,
  FlowContext,
  Handlers,
  StepConfig,
} from './types'

type StepDefinition<TConfig extends StepConfig> = {
  config: TConfig
  handler: Handlers<TConfig>
}

type InferHandlerInput<TConfig extends StepConfig> = TConfig extends StepConfig
  ? Parameters<Handlers<TConfig>>[0]
  : never

type TriggerHandlers<TConfig extends StepConfig> = {
  queue?: (
    input: ExtractQueueInput<InferHandlerInput<TConfig>>,
    ctx: Omit<FlowContext<any, any>, 'match'>,
  ) => Promise<void>
  http?: (
    request: ExtractApiInput<InferHandlerInput<TConfig>>,
    ctx: Omit<FlowContext<any, any>, 'match'>,
  ) => Promise<any>
  cron?: (ctx: Omit<FlowContext<any, any>, 'match'>) => Promise<void>
  state?: (
    input: ExtractStateInput<InferHandlerInput<TConfig>>,
    ctx: Omit<FlowContext<any, any>, 'match'>,
  ) => Promise<any>
  stream?: (
    input: ExtractStreamInput<InferHandlerInput<TConfig>>,
    ctx: Omit<FlowContext<any, any>, 'match'>,
  ) => Promise<any>
}

type MultiTriggerStepBuilder<TConfig extends StepConfig> = {
  config: TConfig
  handlers: (handlers: TriggerHandlers<TConfig>) => StepDefinition<TConfig>

  // Chainable methods
  onQueue: (
    handler: TriggerHandlers<TConfig>['queue'],
  ) => MultiTriggerStepBuilder<TConfig> & { handlers: () => StepDefinition<TConfig> }
  onHttp: (
    handler: TriggerHandlers<TConfig>['http'],
  ) => MultiTriggerStepBuilder<TConfig> & { handlers: () => StepDefinition<TConfig> }
  onCron: (
    handler: TriggerHandlers<TConfig>['cron'],
  ) => MultiTriggerStepBuilder<TConfig> & { handlers: () => StepDefinition<TConfig> }
  onState: (
    handler: TriggerHandlers<TConfig>['state'],
  ) => MultiTriggerStepBuilder<TConfig> & { handlers: () => StepDefinition<TConfig> }
  onStream: (
    handler: TriggerHandlers<TConfig>['stream'],
  ) => MultiTriggerStepBuilder<TConfig> & { handlers: () => StepDefinition<TConfig> }
}

export function multiTriggerStep<const TConfig extends StepConfig>(config: TConfig): MultiTriggerStepBuilder<TConfig> {
  const collectedHandlers: Partial<TriggerHandlers<TConfig>> = {}

  const createUnifiedHandler = (): Handlers<TConfig> => {
    return (async (input: any, ctx: any) => {
      if (ctx.trigger.type === 'queue' && collectedHandlers.queue) {
        return collectedHandlers.queue(input, ctx)
      }
      if (ctx.trigger.type === 'http' && collectedHandlers.http) {
        return collectedHandlers.http(input, ctx)
      }
      if (ctx.trigger.type === 'cron' && collectedHandlers.cron) {
        return collectedHandlers.cron(ctx)
      }
      if (ctx.trigger.type === 'state' && collectedHandlers.state) {
        return collectedHandlers.state(input, ctx)
      }
      if (ctx.trigger.type === 'stream' && collectedHandlers.stream) {
        return collectedHandlers.stream(input, ctx)
      }

      ctx.logger.warn(`No handler defined for trigger type: ${ctx.trigger.type}`, {
        availableHandlers: Object.keys(collectedHandlers),
        triggerType: ctx.trigger.type,
      })

      throw new Error(
        `No handler defined for trigger type: ${ctx.trigger.type}. Available handlers: ${Object.keys(collectedHandlers).join(', ')}`,
      )
    }) as Handlers<TConfig>
  }

  const builder: any = {
    config,

    onQueue(handler: TriggerHandlers<TConfig>['queue']) {
      collectedHandlers.queue = handler
      return {
        ...builder,
        handlers: () => ({ config, handler: createUnifiedHandler() }),
      }
    },

    onHttp(handler: TriggerHandlers<TConfig>['http']) {
      collectedHandlers.http = handler
      return {
        ...builder,
        handlers: () => ({ config, handler: createUnifiedHandler() }),
      }
    },

    onCron(handler: TriggerHandlers<TConfig>['cron']) {
      collectedHandlers.cron = handler
      return {
        ...builder,
        handlers: () => ({ config, handler: createUnifiedHandler() }),
      }
    },

    onState(handler: TriggerHandlers<TConfig>['state']) {
      collectedHandlers.state = handler
      return {
        ...builder,
        handlers: () => ({ config, handler: createUnifiedHandler() }),
      }
    },

    onStream(handler: TriggerHandlers<TConfig>['stream']) {
      collectedHandlers.stream = handler
      return {
        ...builder,
        handlers: () => ({ config, handler: createUnifiedHandler() }),
      }
    },

    handlers(h?: TriggerHandlers<TConfig>) {
      if (h) {
        Object.assign(collectedHandlers, h)
      }
      return { config, handler: createUnifiedHandler() }
    },
  }

  return builder
}

File: frameworks/motia/motia-py/packages/motia/src/motia/multi_trigger.py
"""Multi-trigger step builder for Motia steps with multiple trigger types."""

from __future__ import annotations

from typing import Any, Awaitable, Callable

from .step import StepDefinition, StepHandler
from .types import StepConfig

TriggerHandlers = dict[str, Callable[..., Awaitable[Any]]]


class MultiTriggerStepBuilder:
    """Builder for steps with multiple trigger types.

    Supports both dict-style and chainable handler registration:

        # Dict style:
        multi_trigger_step(config).handlers({
            "queue": queue_handler,
            "http": http_handler,
        })

        # Chainable style:
        multi_trigger_step(config)
            .on_queue(queue_handler)
            .on_http(http_handler)
            .handlers()
    """

    def __init__(self, config: StepConfig) -> None:
        self.config = config
        self._handlers: TriggerHandlers = {}

    def _create_unified_handler(self) -> StepHandler:
        """Create a unified handler that dispatches based on trigger type."""
        handlers = dict(self._handlers)

        async def unified_handler(input_data: Any, ctx: Any) -> Any:
            trigger_type = ctx.trigger.type
            if trigger_type == "queue" and "queue" in handlers:
                return await handlers["queue"](input_data, ctx)
            if trigger_type == "http" and "http" in handlers:
                return await handlers["http"](input_data, ctx)
            if trigger_type == "cron" and "cron" in handlers:
                return await handlers["cron"](ctx)
            if trigger_type == "state" and "state" in handlers:
                return await handlers["state"](input_data, ctx)
            if trigger_type == "stream" and "stream" in handlers:
                return await handlers["stream"](input_data, ctx)

            raise RuntimeError(
                f"No handler defined for trigger type: {trigger_type}. Available handlers: {', '.join(handlers.keys())}"
            )

        return unified_handler

    def on_queue(self, handler: Callable[..., Awaitable[Any]]) -> MultiTriggerStepBuilder:
        """Register a handler for queue triggers."""
        self._handlers["queue"] = handler
        return self

    def on_http(self, handler: Callable[..., Awaitable[Any]]) -> MultiTriggerStepBuilder:
        """Register a handler for HTTP triggers."""
        self._handlers["http"] = handler
        return self

    def on_cron(self, handler: Callable[..., Awaitable[Any]]) -> MultiTriggerStepBuilder:
        """Register a handler for cron triggers."""
        self._handlers["cron"] = handler
        return self

    def on_state(self, handler: Callable[..., Awaitable[Any]]) -> MultiTriggerStepBuilder:
        """Register a handler for state triggers."""
        self._handlers["state"] = handler
        return self

    def on_stream(self, handler: Callable[..., Awaitable[Any]]) -> MultiTriggerStepBuilder:
        """Register a handler for stream triggers."""
        self._handlers["stream"] = handler
        return self

    def handlers(self, handlers: TriggerHandlers | None = None) -> StepDefinition:
        """Finalize the builder and create a StepDefinition.

        Args:
            handlers: Optional dict of trigger type -> handler mappings.
                     If provided, merges with any previously registered handlers.
                     If omitted, uses only the handlers set via on_*() methods.

        Returns:
            StepDefinition with the unified handler.
        """
        if handlers:
            self._handlers.update(handlers)
        return StepDefinition(config=self.config, handler=self._create_unified_handler())


def multi_trigger_step(
    config: StepConfig | dict[str, Any],
) -> MultiTriggerStepBuilder:
    """Create a multi-trigger step builder.

    Usage:
        # Dict-style handlers:
        my_step = multi_trigger_step(config).handlers({
            "queue": queue_handler,
            "http": http_handler,
        })

        # Chainable handlers:
        my_step = (
            multi_trigger_step(config)
            .on_queue(queue_handler)
            .on_http(http_handler)
            .handlers()
        )

    Args:
        config: Step configuration dict or StepConfig instance.

    Returns:
        MultiTriggerStepBuilder for registering handlers.
    """
    if not isinstance(config, StepConfig):
        config = StepConfig.model_validate(config)
    return MultiTriggerStepBuilder(config)
